File: src/main/java/com/company/erp/features/chats/service/MessageService.java
package com.company.erp.features.chats.service;

import com.company.erp.core.database.PageQuery;
import com.company.erp.core.exceptions.BadRequestException;
import com.company.erp.core.exceptions.ForbiddenException;
import com.company.erp.core.exceptions.NotFoundException;
import com.company.erp.features.chats.dto.EditMessageRequest;
import com.company.erp.features.chats.dto.SendMessageRequest;
import com.company.erp.features.chats.entity.Conversation;
import com.company.erp.features.chats.entity.Message;
import com.company.erp.features.chats.entity.MessageReaction;
import com.company.erp.features.chats.entity.MessageReactionId;
import com.company.erp.features.chats.entity.MessageType;
import com.company.erp.features.chats.repository.ConversationRepository;
import com.company.erp.features.chats.repository.MessageReactionRepository;
import com.company.erp.features.chats.repository.MessageRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Set;

@Service
@Transactional
public class MessageService {

    /** Edit window mirrors the mobile guide (Slice 4.8 — 15 minutes). */
    private static final Duration EDIT_WINDOW = Duration.ofMinutes(15);

    private final MessageRepository messages;
    private final MessageReactionRepository reactions;
    private final ConversationRepository conversations;
    private final ConversationService conversationService;

    public MessageService(MessageRepository messages,
                          MessageReactionRepository reactions,
                          ConversationRepository conversations,
                          ConversationService conversationService) {
        this.messages = messages;
        this.reactions = reactions;
        this.conversations = conversations;
        this.conversationService = conversationService;
    }

    @Transactional(readOnly = true)
    public Page<Message> history(Long convId, Long userId, PageQuery query) {
        conversationService.requireMember(convId, userId);
        return messages.findByConversationIdAndDeletedAtIsNullOrderByCreatedAtDesc(convId,
                query.toPageable(Set.of("createdAt"), Sort.by(Sort.Direction.DESC, "createdAt")));
    }

    @Transactional(readOnly = true)
    public Page<Message> search(Long convId, Long userId, String q, PageQuery query) {
        conversationService.requireMember(convId, userId);
        if (q == null || q.isBlank()) {
            return Page.empty(query.toPageable(Set.of("createdAt"),
                    Sort.by(Sort.Direction.DESC, "createdAt")));
        }
        return messages.searchInConversation(convId, q,
                query.toPageable(Set.of("createdAt"), Sort.by(Sort.Direction.DESC, "createdAt")));
    }

    @Transactional(readOnly = true)
    public Message getById(Long messageId) {
        return messages.findById(messageId)
                .orElseThrow(() -> new NotFoundException("Message not found"));
    }

    public Message send(Long convId, Long senderId, SendMessageRequest req) {
        conversationService.requireMember(convId, senderId);
        validateBody(req);

        Message m = new Message();
        m.setConversationId(convId);
        m.setSenderId(senderId);
        m.setType(req.type());
        m.setBody(req.body());
        m.setAttachmentUrl(req.attachmentUrl());
        m.setAttachmentContentType(req.attachmentContentType());
        m.setAttachmentSizeBytes(req.attachmentSizeBytes());
        m.setDurationSeconds(req.durationSeconds());
        if (req.replyToMessageId() != null) {
            Message parent = messages.findById(req.replyToMessageId())
                    .orElseThrow(() -> new NotFoundException("Reply target not found"));
            if (!parent.getConversationId().equals(convId)) {
                throw new BadRequestException("Reply must reference a message in the same conversation");
            }
            m.setReplyToMessageId(parent.getId());
        }
        messages.save(m);

        Conversation c = conversations.findById(convId).orElseThrow();
        c.setLastMessageId(m.getId());
        c.setLastMessageAt(m.getCreatedAt());
        return m;
    }

    public Message edit(Long messageId, Long actorId, EditMessageRequest req) {
        Message m = getById(messageId);
        if (!m.getSenderId().equals(actorId)) throw new ForbiddenException("Can only edit own messages");
        if (m.isDeleted())                    throw new BadRequestException("Cannot edit a deleted message");
        if (m.getType() != MessageType.TEXT)  throw new BadRequestException("Only TEXT messages are editable");
        if (Duration.between(m.getCreatedAt(), Instant.now()).compareTo(EDIT_WINDOW) > 0) {
            throw new BadRequestException("Edit window has expired");
        }
        m.setBody(req.body());
        m.setEditedAt(Instant.now());
        return m;
    }

    public Message delete(Long messageId, Long actorId) {
        Message m = getById(messageId);
        if (!m.getSenderId().equals(actorId)) throw new ForbiddenException("Can only delete own messages");
        if (m.isDeleted()) return m;
        m.setDeletedAt(Instant.now());
        return m;
    }

    /** Toggle a single emoji for the user on a message. Returns the resulting list. */
    public List<MessageReaction> toggleReaction(Long messageId, Long userId, String emoji) {
        Message m = getById(messageId);
        if (m.isDeleted()) throw new BadRequestException("Cannot react to a deleted message");
        conversationService.requireMember(m.getConversationId(), userId);

        boolean has = reactions.existsById_MessageIdAndId_UserIdAndId_Emoji(messageId, userId, emoji);
        if (has) {
            reactions.deleteById_MessageIdAndId_UserIdAndId_Emoji(messageId, userId, emoji);
        } else {
            MessageReaction r = new MessageReaction();
            r.setId(new MessageReactionId(messageId, userId, emoji));
            reactions.save(r);
        }
        return reactions.findById_MessageId(messageId);
    }

    public List<MessageReaction> reactionsFor(Long messageId) {
        return reactions.findById_MessageId(messageId);
    }

    private static void validateBody(SendMessageRequest req) {
        switch (req.type()) {
            case TEXT -> {
                if (req.body() == null || req.body().isBlank()) {
                    throw new BadRequestException("TEXT message requires non-empty body");
                }
            }
            case IMAGE, FILE -> {
                if (req.attachmentUrl() == null || req.attachmentUrl().isBlank()) {
                    throw new BadRequestException(req.type() + " message requires attachmentUrl");
                }
            }
            case VOICE -> {
                if (req.attachmentUrl() == null || req.durationSeconds() == null) {
                    throw new BadRequestException("VOICE message requires attachmentUrl and durationSeconds");
                }
            }
        }
    }
}

File: app/Features/Chat/Services/MessageService.php
<?php

namespace App\Features\Chat\Services;

use App\Features\Chat\Models\ChatConversation;
use App\Features\Chat\Models\ChatMessage;
use App\Features\Chat\Models\MessageReaction;
use App\Features\Chat\Models\MessageType;
use App\Support\Exceptions\BadRequestException;
use App\Support\Exceptions\ForbiddenException;
use App\Support\Exceptions\NotFoundException;
use App\Support\Pagination\PageQuery;
use Illuminate\Contracts\Pagination\LengthAwarePaginator;
use Illuminate\Database\Eloquent\Collection;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\DB;

/**
 * Port of the Spring MessageService.
 */
class MessageService
{
    /** Edit window mirrors the mobile guide (Slice 4.8 — 15 minutes). */
    private const EDIT_WINDOW_SECONDS = 900;

    /** @var ConversationService */
    private $conversations;

    public function __construct(ConversationService $conversations)
    {
        $this->conversations = $conversations;
    }

    public function history(int $convId, int $userId, PageQuery $query): LengthAwarePaginator
    {
        $this->conversations->requireMember($convId, $userId);
        return ChatMessage::query()
            ->where('conversation_id', $convId)
            ->whereNull('deleted_at')
            ->orderBy('created_at', 'desc')
            ->orderBy('id', 'desc')
            ->paginate($query->pageSize, ['*'], 'page', $query->page);
    }

    public function search(int $convId, int $userId, ?string $q, PageQuery $query): LengthAwarePaginator
    {
        $this->conversations->requireMember($convId, $userId);

        $builder = ChatMessage::query()
            ->where('conversation_id', $convId)
            ->whereNull('deleted_at');

        if ($q === null || trim($q) === '') {
            // Match Spring's Page.empty(): a query that returns nothing.
            $builder->whereRaw('1 = 0');
        } else {
            $like = '%'.strtolower($q).'%';
            $builder->whereRaw('LOWER(body) LIKE ?', [$like]);
        }

        return $builder
            ->orderBy('created_at', 'desc')
            ->orderBy('id', 'desc')
            ->paginate($query->pageSize, ['*'], 'page', $query->page);
    }

    public function getById(int $messageId): ChatMessage
    {
        $m = ChatMessage::query()->find($messageId);
        if (! $m) {
            throw new NotFoundException('Message not found');
        }
        return $m;
    }

    /**
     * @param array<string, mixed> $data
     */
    public function send(int $convId, int $senderId, array $data): ChatMessage
    {
        $this->conversations->requireMember($convId, $senderId);
        $this->validateBody($data);

        return DB::transaction(function () use ($convId, $senderId, $data) {
            $m = new ChatMessage();
            $m->conversation_id = $convId;
            $m->sender_id = $senderId;
            $m->type = $data['type'];
            $m->body = $data['body'] ?? null;
            $m->attachment_url = $data['attachmentUrl'] ?? null;
            $m->attachment_content_type = $data['attachmentContentType'] ?? null;
            $m->attachment_size_bytes = $data['attachmentSizeBytes'] ?? null;
            $m->duration_seconds = $data['durationSeconds'] ?? null;

            if (! empty($data['replyToMessageId'])) {
                $parent = ChatMessage::query()->find($data['replyToMessageId']);
                if (! $parent) {
                    throw new NotFoundException('Reply target not found');
                }
                if ((int) $parent->conversation_id !== $convId) {
                    throw new BadRequestException('Reply must reference a message in the same conversation');
                }
                $m->reply_to_message_id = $parent->id;
            }
            $m->save();

            $c = ChatConversation::query()->find($convId);
            if ($c) {
                $c->last_message_id = $m->id;
                $c->last_message_at = $m->created_at;
                $c->save();
            }

            return $m;
        });
    }

    /**
     * @param array<string, mixed> $data  body
     */
    public function edit(int $messageId, int $actorId, array $data): ChatMessage
    {
        return DB::transaction(function () use ($messageId, $actorId, $data) {
            $m = $this->getById($messageId);
            if ((int) $m->sender_id !== $actorId) {
                throw new ForbiddenException('Can only edit own messages');
            }
            if ($m->isDeleted()) {
                throw new BadRequestException('Cannot edit a deleted message');
            }
            if ($m->type !== MessageType::TEXT) {
                throw new BadRequestException('Only TEXT messages are editable');
            }
            if ($m->created_at !== null && $m->created_at->diffInSeconds(Carbon::now()) > self::EDIT_WINDOW_SECONDS) {
                throw new BadRequestException('Edit window has expired');
            }
            $m->body = $data['body'];
            $m->edited_at = Carbon::now();
            $m->save();
            return $m;
        });
    }

    public function delete(int $messageId, int $actorId): ChatMessage
    {
        return DB::transaction(function () use ($messageId, $actorId) {
            $m = $this->getById($messageId);
            if ((int) $m->sender_id !== $actorId) {
                throw new ForbiddenException('Can only delete own messages');
            }
            if ($m->isDeleted()) {
                return $m;
            }
            $m->deleted_at = Carbon::now();
            $m->save();
            return $m;
        });
    }

    /**
     * Toggle a single emoji for the user on a message. Returns the resulting list.
     *
     * @return Collection<int, MessageReaction>
     */
    public function toggleReaction(int $messageId, int $userId, string $emoji): Collection
    {
        return DB::transaction(function () use ($messageId, $userId, $emoji) {
            $m = $this->getById($messageId);
            if ($m->isDeleted()) {
                throw new BadRequestException('Cannot react to a deleted message');
            }
            $this->conversations->requireMember((int) $m->conversation_id, $userId);

            $exists = MessageReaction::query()
                ->where('message_id', $messageId)
                ->where('user_id', $userId)
                ->where('emoji', $emoji)
                ->exists();

            if ($exists) {
                MessageReaction::query()
                    ->where('message_id', $messageId)
                    ->where('user_id', $userId)
                    ->where('emoji', $emoji)
                    ->delete();
            } else {
                $r = new MessageReaction();
                $r->message_id = $messageId;
                $r->user_id = $userId;
                $r->emoji = $emoji;
                $r->created_at = Carbon::now();
                $r->save();
            }

            return $this->reactionsFor($messageId);
        });
    }

    /**
     * @return Collection<int, MessageReaction>
     */
    public function reactionsFor(int $messageId): Collection
    {
        return MessageReaction::query()->where('message_id', $messageId)->get();
    }

    /**
     * @param array<string, mixed> $data
     */
    private function validateBody(array $data): void
    {
        $type = $data['type'];
        $body = $data['body'] ?? null;
        $attachmentUrl = $data['attachmentUrl'] ?? null;
        $durationSeconds = $data['durationSeconds'] ?? null;

        if ($type === MessageType::TEXT) {
            if ($body === null || trim($body) === '') {
                throw new BadRequestException('TEXT message requires non-empty body');
            }
        } elseif ($type === MessageType::IMAGE || $type === MessageType::FILE) {
            if ($attachmentUrl === null || trim($attachmentUrl) === '') {
                throw new BadRequestException($type.' message requires attachmentUrl');
            }
        } elseif ($type === MessageType::VOICE) {
            if ($attachmentUrl === null || $durationSeconds === null) {
                throw new BadRequestException('VOICE message requires attachmentUrl and durationSeconds');
            }
        }
    }
}
